File: tokenizer.py
from typing import List

# TODO:
# 1. Can add unknown token for foreign chars.
# 2. C++/Rust for faster implementation
# 3. Pre load a-z and A-Z
# 4. Trie implementation for single parse through text to encode


class BPETokenizer:
    def __init__(self):
        self.itos = {}
        self.stoi = {}
        self.vocab_size = 0
        self.max_key = 0
        self.merges = {}
    
    def train(self, text_corpus: str, vocab_size: int, verbose: bool = False):
        """
        Takes in the corpus of text to train on and the vocabulary size.
        
        Input: 
            text_corpus: str - The corpus of text to train the tokenizer on.

            vocab_size: int - The size of the vocabulary of the tokenizer.
        """
        self.vocab_size = vocab_size
        text_corpus_set = set(text_corpus)
        text_corpus_unique = sorted(text_corpus_set)

        assert len(text_corpus_unique) <= vocab_size, f"vocab_size ({vocab_size}) must be greater than or equal to the number of unique chars ({len(text_corpus_unique)}) in the text corpus"

        for ch in text_corpus_unique:
            self.stoi[ch] = self.max_key
            self.itos[self.max_key] = ch
            self.max_key += 1

        text_idx = [self.stoi[ch] for ch in text_corpus]

        milestones = [0] + [int(vocab_size * frac) - 1 for frac in (0.1 * i for i in range(2, 11, 2))]

        while self.max_key < vocab_size:
            if verbose:
                if self.max_key in milestones:
                    percent_trained = int(round((100 * (self.max_key + 1) / vocab_size)))
                    print(f"Training progress: {percent_trained}%")

            pair_count = {}
            for i1, i2 in zip(text_idx, text_idx[1:]):
                pair_count[(i1, i2)] = pair_count.get((i1, i2), 0) + 1

            max_pair = max(pair_count, key=lambda pair: pair_count[pair])
            new_text_idx = []

            i = 0
            while i < len(text_idx)-1:
                if (text_idx[i], text_idx[i+1]) == max_pair:
                    new_text_idx.append(self.max_key)
                    i += 2
                else:
                    new_text_idx.append(text_idx[i])
                    i += 1
            
            merged_pair = self.itos[max_pair[0]] + self.itos[max_pair[1]]
            self.itos[self.max_key] = merged_pair
            self.stoi[merged_pair] = self.max_key

            self.merges[max_pair] = self.max_key
            self.max_key += 1
            text_idx = new_text_idx

        self.final_vocab = sorted(self.stoi.items(), key=lambda item: len(item[0]), reverse=True)
        print("Tokenizer successfully trained!")
        return
    
    def encode(self, text: str):
        """
        Function to encode a given string.

        Input:
            text: str - The text (string) which is to be encoded.
        """
        encoded_idx = []
        rem_text = text

        while rem_text:
            for substr, subkey in self.final_vocab:
                if rem_text.startswith(substr):
                    encoded_idx.append(subkey)
                    rem_text = rem_text[len(substr):]
                    break
        
        return encoded_idx

    def decode(self, idx: List[int]):
        """
        Function to decode a given list of encoded indices.

        Input:
            idx: List[int] - The list of encoded indices (encoding) which is to be decoded.
        """
        return "".join(self.itos[i] for i in idx)

File: tokenizer.cpp
#include "tokenizer.h"

#include <iostream>
#include <sstream>
#include <stdexcept>
#include <cmath>
#include <algorithm>
#include <set>

BPETokenizer::BPETokenizer() : vocab_size(0), max_key(0) {

}


/**
 * @brief Trains the tokenizer on the provided text corpus.
 * 
 * This function builds a vocabulary up to the given size by applying
 * Byte Pair Encoding (BPE) merges on the text.
 * 
 * @param text_corpus The input text corpus for training.
 * @param vocab_size Desired vocabulary size after training.
 * @param verbose Enable verbose output for debugging or progress tracking.
 */
void BPETokenizer::train(std::string text_corpus, int vocab_size, bool verbose) {
    this->vocab_size = vocab_size;
    std::set<char> text_corpus_unique(text_corpus.begin(), text_corpus.end());

    if (text_corpus_unique.size() > vocab_size) {
        std::ostringstream oss;
        oss << "vocab_size (" << vocab_size << ") must be greater than or equal to the number of unique chars (" << text_corpus_unique.size() << ") in the text corpus";
        throw std::runtime_error(oss.str());
    }

    for (char ch : text_corpus_unique) {
        std::string str(1, ch);
        stoi[str] = max_key;
        itos[max_key] = str;
        max_key++;
    }

    std::vector<int> text_idx;
    for (char ch : text_corpus) {
        std::string str(1, ch);
        text_idx.push_back(stoi[str]);
    }

    std::set<int> milestones{0};
    for (int i = 2; i < 11; i += 2) {
        float frac = 0.1 * i;
        int milestone = static_cast<int>(vocab_size * frac) - 1;
        milestones.insert(milestone);
    }

    while (max_key < vocab_size) {
        if (verbose) {
            if (milestones.find(max_key) != milestones.end()) {
                double progress = static_cast<double>(max_key + 1) / vocab_size;
                int percent_trained = static_cast<int>(std::round(progress * 100));
                std::cout << "Training progress " << std::to_string(percent_trained) << "%\n";
            }
        }

        std::unordered_map<std::pair<int, int>, int, pair_hash> pair_count;
        for (int i = 0; i < text_idx.size()-1; i++) {
            pair_count[std::make_pair(text_idx[i], text_idx[i+1])]++;
        }

        if (pair_count.empty()) break;

        auto max_pair = std::max_element(pair_count.begin(), pair_count.end(),
            [](const auto& a, const auto& b) {
                return a.second < b.second;
            })->first;

        std::vector<int> new_text_idx;

        int i{0};
        while (i < text_idx.size()-1) {
            if (std::pair<int, int>{text_idx[i], text_idx[i+1]} == max_pair) {
                new_text_idx.push_back(max_key);
                i += 2;
            } else {
                new_text_idx.push_back(text_idx[i]);
                i++;
            }
        }

        std::string merged_pair = itos[max_pair.first] + itos[max_pair.second];
        itos[max_key] = merged_pair;
        stoi[merged_pair] = max_key;

        merges[max_pair] = max_key;
        max_key++;
        text_idx = std::move(new_text_idx);
    }

    final_vocab.clear();
    for (const auto& [str, id] : stoi) {
        final_vocab.emplace_back(str, id);
    }
    std::sort(final_vocab.begin(), final_vocab.end(),
        [](const auto& a, const auto& b) {
            return a.first.length() > b.first.length();
        });
    
    if (verbose) {
        std::cout << "Tokenizer successfully trained! Final vocab size: " << stoi.size() << "\n";
    }
}

/**
 * @brief Function to encode a given string.
 * 
 * @param text The text (string) which is to be encoded.
 * 
 * @return A vector of encoded integers.
*/
std::vector<int> BPETokenizer::encode(std::string text) {
    std::vector<int> encoded_idx;
    size_t pos = 0;

    while (pos < text.size()) {
        bool matched = false;

        for (const auto& [substr, subkey] : this->final_vocab) {
            size_t len = substr.length();

            if (pos + len <= text.length() && text.compare(pos, len, substr) == 0) {
                encoded_idx.push_back(subkey);
                pos += len;
                matched = true;
                break;
            }
        }

        if (!matched) {
            ++pos;
        }
    }
    
    return encoded_idx;
}

/**
 * @brief Function to decode a given vector of encoded indices.
 * 
 * @param idx The vector of encoded indices (encoding) which is to be decoded.
 * 
 * @return The decoded string.
*/
std::string BPETokenizer::decode(std::vector<int> idx)
{
    // add support for unsupported/unknown tokens
    std::string decoded_str;
    for (int i : idx) {
        decoded_str += itos[i];
    }

    return decoded_str;
}
